### PR TITLE
Refresh token logic

### PR DESCRIPTION
**Pull request by fortamt**
develop

## GitHub Board

**Issue link**

[Refresh Token logic #549](https://github.com/ita-social-projects/dokazovi-be/issues/549 )

**Story link**

[-](- )

## Code reviewers
- [ ] @teaFunny 

## Summary of issue

- [x]  create /refreshtoken endpoint
- [x] add refresh token logic

## Summary of change

- [x]  created /refreshtoken endpoint
- [x]  added refresh token repository
- [x] added table for refresh tokens(flyway migrations)
- [x] added RefreshTokenService.class
- [x] added Tests for all
- [x] added links for enviroment variables

### CODE REVIEW

File: src/main/java/com/softserveinc/dokazovi/security/RefreshTokenService.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		long expireTime = appProperties.getAuth().getRefreshTokenExpirationMsec();
+
+		Optional<UserEntity> userEntity = userRepository.findById(userId);
+		if (userEntity.isPresent()) {
+			refreshToken.setUser(userEntity.get());
+			refreshToken.setExpiryDate(Instant.now().plusMillis(expireTime));
+			refreshToken.setToken(UUID.randomUUID().toString());
+		}
+
+		refreshToken = refreshTokenRepository.save(refreshToken);
```
Comment (V-Kaidash) [inline src/main/java/com/softserveinc/dokazovi/security/RefreshTokenService.java:39]: Does something changes after 
`refreshTokenRepository.save(refreshToken);`
Why are you reassigning refreshToken variable?

File: src/main/java/com/softserveinc/dokazovi/security/RefreshTokenService.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	public RefreshToken verifyExpiration(RefreshToken token) {
+		if (token.getExpiryDate().compareTo(Instant.now()) < 0) {
+			refreshTokenRepository.delete(token);
+			throw new TokenRefreshException(token.getToken(),
+					"Refresh token was expired. Please make a new signin request");
+		}
+		return token;
+	}
+
+
```
Comment (V-Kaidash) [inline src/main/java/com/softserveinc/dokazovi/security/RefreshTokenService.java:53]: Please remove empty lines

File: src/main/java/com/softserveinc/dokazovi/security/TokenProvider.java [hunk 3]
```diff
@@ -43,6 +47,7 @@ public String createToken(Authentication authentication) {
 				.compact();
 	}
 
+
```
Comment (V-Kaidash) [inline src/main/java/com/softserveinc/dokazovi/security/TokenProvider.java:50]: Please remove empty lines

File: src/main/resources/application.properties [hunk 4]
```diff
@@ -36,3 +36,9 @@ spring.mail.properties.mail.smtp.auth=true
 spring.mail.properties.mail.smtp.starttls.enable=true
 spring.mail.properties.mail.smtp.starttls.required=true
 spring.mail.properties.mail.smtp.timeout=8000
+
+#-------------------------
+# Token Expiration Settings
+#-------------------------
+app.auth.tokenExpirationMsec = ${TOKEN_EXPIRATION_MSEC}
+app.auth.refreshTokenExpirationMsec= ${REFRESH_TOKEN_EXPIRATION_MSEC}
```
Comment (V-Kaidash) [inline src/main/resources/application.properties:44]: What happened if these two parameters not set?
Does it make sense to set some default values?

File: src/main/java/com/softserveinc/dokazovi/config/AppProperties.java [hunk 5]
```diff
@@ -1,55 +1,71 @@
...
-    }
-
-    public OAuth2 getOauth2() {
-        return oauth2;
-    }
+
+	private final Auth auth = new Auth();
+	private final OAuth2 oauth2 = new OAuth2();
+
+	public static class Auth {
```
Comment (V-Kaidash) [inline src/main/java/com/softserveinc/dokazovi/config/AppProperties.java:15]: Please check your IDE settings for indentation and tabs and adjust your code accordingly. Also check other files you have edited. IDE settings should be something like this 
image

File: src/test/java/com/softserveinc/dokazovi/security/RefreshTokenServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	@Test
+	void verifyExpirationFalse() {
+		RefreshToken refreshToken = new RefreshToken();
+		refreshToken.setId(1);
+		refreshToken.setUser(user);
+		refreshToken.setExpiryDate(Instant.now().minusMillis(60000L));
+		refreshToken.setToken("950c9760-805e-449c-a966-2d0d5ebd86f4");
+		assertThrows(TokenRefreshException.class, () -> refreshTokenService.verifyExpiration(refreshToken));
+	}
+
```
Comment (V-Kaidash) [inline src/test/java/com/softserveinc/dokazovi/security/RefreshTokenServiceTest.java:84]: Please remove empty lines & fix indentation (see IDE settings above)

**Review by V-Kaidash — CHANGES_REQUESTED** [review]

**Review by V-Kaidash — APPROVED** [review]
Refresh token proof/demo
https://drive.google.com/file/d/1KBhr29pPqhQ5N0Asw3RhYo1kZlmL02fS/view?usp=sharing

All looks good for me.
Let's wait on the final review from Alina before merging.

**Review by teaFunny — APPROVED** [review]

**Pull request merged**